app/controllers/api/v1/users_controller.rb

```diff
@@ -47,17 +47,11 @@ def me
   end
 
   def update
-    [].tap do |update_email_user_ids|
-
-      super do |user|
-        if user.email_changed?
-          update_email_user_ids << user.id
-          user.update_attribute(:valid_email, true)
-        end
-      end
-
-      update_email_user_ids.each do |user_id|
-        UserInfoChangedMailerWorker.perform_async(user_id, "email")
+    prev_email = user.email
+    super do |user|
+      if user.email_changed?
+        user.update(valid_email: true)
+        UserInfoChangedMailerWorker.perform_async(user.id, 'email', prev_email)
```

Comment:
I think this will reintroduce the race condition that the tap/array-with-a-single-id thing seeks to avoid. The save in `update` happens after the block is yielded, so when the worker is called before the `super` block is finished, the user hasn't been updated in the db and the email will go to the previous address. 

Moving the worker call outside of the super block should be sufficient, assuming that `user.email_changed?` is still accurate after the save. It's possible that "changed?" info is lost after the transaction, though, and perhaps that's why the `update_email_user_ids.each` exists. Either way, it's gotta be outside the super block in order to ensure that the new address is saved while you pass along the old one.

Reply:
I get now. The user.email_changed? is not very accurate outside the save so i'm passing it through a variable now